File: global localization/live/cpp/src/gnss_reader/GPSDReader.cpp
#include "gnss_reader/GPSDReader.hpp"

GPSDReader::GPSDReader() {

}

GPSDReader::~GPSDReader() {
    continue_to_grab = false;
    grab_gnss_data.join();
#ifdef GPSD_FOUND

#else
    std::cerr << "[library not found] GPSD library was not found ... please install it before using this sample" << std::endl;
#endif
}

void GPSDReader::initialize() {
    grab_gnss_data = std::thread(&GPSDReader::grabGNSSData, this);
#ifdef GPSD_FOUND
    gnss_getter.reset(new gpsmm("localhost", DEFAULT_GPSD_PORT));
    if (gnss_getter->stream(WATCH_ENABLE | WATCH_JSON) == nullptr) {
        std::cerr << "No GPSD running .. exit" << std::endl;
        exit(EXIT_FAILURE);
    }
    std::cout << "Successfully opened GPS stream" << std::endl;
    std::cout << "Waiting for GNSS fix" << std::endl;

    bool received_fix = false;
    struct gps_data_t *gpsd_data;
    while (!received_fix) {
        if (!gnss_getter->waiting(0))
            continue;
        if ((gpsd_data = gnss_getter->read()) == NULL) {
            std::cerr << "[GNSS] read error ... exit program" << std::endl;
            exit(EXIT_FAILURE);
        }
        if (gpsd_data->fix.mode >= MODE_2D)
            received_fix = true;
    }
    std::cout << "Fix found !!!" << std::endl;
    is_initialized_mtx.lock();
    is_initialized = true;
    is_initialized_mtx.unlock();
#else
    std::cerr << "[library not found] GPSD library was not found ... please install it before using this sample" << std::endl;
#endif
}

sl::GNSSData GPSDReader::getNextGNSSValue() {
#ifdef GPSD_FOUND
    // 0. Check if GNSS is initialized:
    // 1. Get GNSS datas:
    struct gps_data_t *gpsd_data;
    while ((gpsd_data = gnss_getter->read()) == NULL)
        ;
    if (gpsd_data->fix.mode >= MODE_2D) {
        int nb_low_snr = 0;
        for (int i = 0; i < gpsd_data->satellites_visible; i++) {
            satellite_t &satellite = gpsd_data->skyview[i];
            if (satellite.used && satellite.ss < 16) nb_low_snr++;
        }
        if (nb_low_snr > 0) std::cout << "[Warning] Low SNR (<16) on " << nb_low_snr << " satellite(s) (using " << gpsd_data->satellites_used << " out of " << gpsd_data->satellites_visible << " visible)" << std::endl;

        sl::GNSSData current_gnss_data;
        // Fill out coordinates:
        current_gnss_data.setCoordinates(gpsd_data->fix.latitude, gpsd_data->fix.longitude, gpsd_data->fix.altMSL, false);
        // Fill out default standard deviation:
        current_gnss_data.longitude_std = current_gnss_data.latitude_std = 0.001f;
        current_gnss_data.altitude_std = 1.f;
        // Fill out covariance [must be not null]
        std::array<double, 9> position_covariance;
        position_covariance[0] = gpsd_data->fix.eph * gpsd_data->fix.eph;
        position_covariance[1 * 3 + 1] = gpsd_data->fix.eph * gpsd_data->fix.eph;
        position_covariance[2 * 3 + 2] = gpsd_data->fix.epv * gpsd_data->fix.epv;
        current_gnss_data.position_covariance = position_covariance;
        // Compute timestamp
        uint64_t current_ts_gps = gpsd_data->fix.time.tv_sec * 1000000;
        uint64_t current_tns_gps = gpsd_data->fix.time.tv_nsec / float(1000);
        auto current_gnss_timestamp = current_ts_gps + current_tns_gps;
        current_gnss_data.ts.setMicroseconds(current_gnss_timestamp);

        int gpsd_mode = gpsd_data->fix.mode;
        sl::GNSS_MODE sl_mode = sl::GNSS_MODE::UNKNOWN;

        switch (gpsd_mode) {
            case 0: // MODE_NOT_SEEN
                sl_mode = sl::GNSS_MODE::UNKNOWN;
                break;
            case 1: // MODE_NO_FIX
                sl_mode = sl::GNSS_MODE::NO_FIX;
                break;
            case 2: // MODE_2D
                sl_mode = sl::GNSS_MODE::FIX_2D;
                break;
            case 3: // MODE_3D
                sl_mode = sl::GNSS_MODE::FIX_3D;
                break;
            default:
                sl_mode = sl::GNSS_MODE::UNKNOWN;
                break;
        }

        int gpsd_status = gpsd_data->fix.status;
        sl::GNSS_STATUS sl_status = sl::GNSS_STATUS::UNKNOWN;

        switch (gpsd_status) {
            case 0: // STATUS_UNK
                sl_status = sl::GNSS_STATUS::UNKNOWN;
                break;
            case 1: // STATUS_GPS
                sl_status = sl::GNSS_STATUS::SINGLE;
                break;
            case 2: // STATUS_DGPS
                sl_status = sl::GNSS_STATUS::DGNSS;
                break;
            case 3: // STATUS_RTK_FIX
                sl_status = sl::GNSS_STATUS::RTK_FIX;
                break;
            case 4: // STATUS_RTK_FLT
                sl_status = sl::GNSS_STATUS::RTK_FLOAT;
                break;
            case 5: // STATUS_DR
                sl_status = sl::GNSS_STATUS::SINGLE;
                break;
            case 6: // STATUS_GNSSDR
                sl_status = sl::GNSS_STATUS::DGNSS;
                break;
            case 7: // STATUS_TIME
                sl_status = sl::GNSS_STATUS::UNKNOWN;
                break;
            case 8: // STATUS_SIM
                sl_status = sl::GNSS_STATUS::UNKNOWN;
                break;
            case 9: // STATUS_PPS_FIX
                sl_status = sl::GNSS_STATUS::SINGLE;
                break;
            default:
                sl_status = sl::GNSS_STATUS::UNKNOWN;
                break;
        }

        current_gnss_data.gnss_status = sl_status;
        current_gnss_data.gnss_mode = sl_mode;

        return current_gnss_data;
    } else {
        std::cout << "Fix lost: reinit GNSS" << std::endl;
        initialize();
        return getNextGNSSValue();
    }
#else
    std::cerr << "[library not found] GPSD library was not found ... please install it before using this sample" << std::endl;
#endif

    return sl::GNSSData();
}

sl::ERROR_CODE GPSDReader::grab(sl::GNSSData & current_data) {
    if (new_data) {
        new_data = false;
        current_data = current_gnss_data;
        return sl::ERROR_CODE::SUCCESS;
    }
    return sl::ERROR_CODE::FAILURE;
}

void GPSDReader::grabGNSSData() {
    while (1) {
        is_initialized_mtx.lock();
        if (is_initialized) {
            is_initialized_mtx.unlock();
            break;
        }
        is_initialized_mtx.unlock();
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
    while (continue_to_grab) {
#ifdef GPSD_FOUND
        current_gnss_data = getNextGNSSValue();
        new_data = true;
#endif
    }

}

File: global localization/live/python/gnss_reader/gpsd_reader.py
import threading
import time
import pyzed.sl as sl
from gpsdclient import GPSDClient


class GPSDReader:
    def __init__(self):
        self.continue_to_grab = True
        self.new_data = False
        self.is_initialized = False
        self.current_gnss_data = None
        self.is_initialized_mtx = threading.Lock()
        self.client = None
        self.gnss_getter = None 

    def initialize(self):
        try : 
            self.client = GPSDClient(host="127.0.0.1")
        except : 
            print("No GPSD running .. exit")
            return -1 
        self.grab_gnss_data = threading.Thread(target=self.grabGNSSData)
        self.grab_gnss_data.start()
        print("Successfully connected to GPSD")
        print("Waiting for GNSS fix")
        received_fix = False
        while not received_fix:
            self.gnss_getter = self.client.dict_stream(convert_datetime=True, filter=["TPV", "SKY"])
            gpsd_data = next(self.gnss_getter)
            if "class" in gpsd_data and gpsd_data["class"] == "TPV" and "mode" in gpsd_data and gpsd_data["mode"] >=2:
                received_fix = True
        print("Fix found !!!")
        with self.is_initialized_mtx:
            self.is_initialized = True
        return 0 

    def getNextGNSSValue(self):
        while self.continue_to_grab :
            gpsd_data = None 
            while gpsd_data is None:
                gpsd_data = next(self.gnss_getter)
            if "class" in gpsd_data and gpsd_data["class"] == "TPV" and "mode" in gpsd_data and gpsd_data["mode"] >=2:                    
                current_gnss_data = sl.GNSSData()
                current_gnss_data.set_coordinates(gpsd_data["lat"], gpsd_data["lon"], gpsd_data["altMSL"], False)
                current_gnss_data.longitude_std =  0.001
                current_gnss_data.latitude_std = 0.001
                current_gnss_data.altitude_std = 1.0

                gpsd_mode = gpsd_data["mode"]
                sl_mode = sl.GNSS_MODE.UNKNOWN

                if gpsd_mode == 0:  # MODE_NOT_SEEN
                    sl_mode = sl.GNSS_MODE.UNKNOWN
                elif gpsd_mode == 1:  # MODE_NO_FIX
                    sl_mode = sl.GNSS_MODE.NO_FIX
                elif gpsd_mode == 2:  # MODE_2D
                    sl_mode = sl.GNSS_MODE.FIX_2D
                elif gpsd_mode == 3:  # MODE_3D
                    sl_mode = sl.GNSS_MODE.FIX_3D

                sl_status = sl.GNSS_STATUS.UNKNOWN
                if 'status' in gpsd_data:
                    gpsd_status = cgpsd_data["status"]
                    if gpsd_status == 0:  # STATUS_UNK
                        sl_status = sl.GNSS_STATUS.UNKNOWN
                    elif gpsd_status == 1:  # STATUS_GPS
                        sl_status = sl.GNSS_STATUS.SINGLE
                    elif gpsd_status == 2:  # STATUS_DGPS
                        sl_status = sl.GNSS_STATUS.DGNSS
                    elif gpsd_status == 3:  # STATUS_RTK_FIX
                        sl_status = sl.GNSS_STATUS.RTK_FIX
                    elif gpsd_status == 4:  # STATUS_RTK_FLT
                        sl_status = sl.GNSS_STATUS.RTK_FLOAT
                    elif gpsd_status == 5:  # STATUS_DR
                        sl_status = sl.GNSS_STATUS.SINGLE
                    elif gpsd_status == 6:  # STATUS_GNSSDR
                        sl_status = sl.GNSS_STATUS.DGNSS
                    elif gpsd_status == 7:  # STATUS_TIME
                        sl_status = sl.GNSS_STATUS.UNKNOWN
                    elif gpsd_status == 8:  # STATUS_SIM
                        sl_status = sl.GNSS_STATUS.UNKNOWN
                    elif gpsd_status == 9:  # STATUS_PPS_FIX
                        sl_status = sl.GNSS_STATUS.SINGLE


                current_gnss_data.gnss_mode = sl_mode.value
                current_gnss_data.gnss_status = sl_status.value
                
                position_covariance = [
                    gpsd_data["eph"] * gpsd_data["eph"],
                    0.0,
                    0.0,
                    0.0,
                    gpsd_data["eph"] * gpsd_data["eph"],
                    0.0,
                    0.0,
                    0.0,
                    gpsd_data["epv"] * gpsd_data["epv"]
                ]
                current_gnss_data.position_covariances = position_covariance
                timestamp_microseconds = int(gpsd_data["time"].timestamp() * 1000000)
                ts = sl.Timestamp()
                ts.set_microseconds(timestamp_microseconds)
                current_gnss_data.ts = ts 
                return current_gnss_data
            elif "class" in gpsd_data and gpsd_data["class"] == "SKY":
                nb_low_snr = 0
                if 'satellites' in gpsd_data:
                    for satellite in gpsd_data['satellites']:
                        if satellite['used'] and satellite['ss'] < 16:
                            nb_low_snr += 1
                    if nb_low_snr > 0:
                        if 'uSat' in gpsd_data and 'nSat' in gpsd_data:
                            print("[Warning] Low SNR (<16) on {} satellite(s) (using {} out of {} visible)".format(nb_low_snr, gpsd_data['uSat'], gpsd_data['nSat']))
                        else:
                            print("[Warning] Low SNR (", nb_low_snr, "< 16 )")
                    return self.getNextGNSSValue()
            else:  
                print("Fix lost: GNSS reinitialization")
                self.initialize()

    def grab(self):
        if self.new_data:
            self.new_data = False
            return sl.ERROR_CODE.SUCCESS, self.current_gnss_data
        return sl.ERROR_CODE.FAILURE, None 

    def grabGNSSData(self):
        while self.continue_to_grab:
            with self.is_initialized_mtx:
                if self.is_initialized:
                    break
            time.sleep(0.001)

        while self.continue_to_grab :
            self.current_gnss_data = self.getNextGNSSValue()
            self.new_data = True

    def stop_thread(self):
        self.continue_to_grab = False
        